test/scale/scale_tester.go

```diff
@@ -0,0 +1,358 @@
+			fiveMinPodsRunning = numberOfRunningPods
+		case timePassed <= 10*time.Minute:
+			tenMinPodsRunning = numberOfRunningPods
+		}
+		time.Sleep(30 * time.Second)
+	}
+	scaleOutCompleted := time.Since(startScaleOut).Minutes()
+
+	startScaleIn := time.Now()
+	newScale := &scale.Scale{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      statefulSet.Statefulset.Name,
+			Namespace: statefulSet.Namespace.Name,
+		},
+		Spec: scale.ScaleSpec{
+			Replicas: int32(0),
+		},
+	}
+	_, _ = client.AppsV1().StatefulSets(statefulSet.Namespace.Name).UpdateScale(context.TODO(), statefulSet.Statefulset.Name, newScale, metav1.UpdateOptions{})
+
+	numberOfRemainingPods := totalNumberOfPods
+	for {
+		numberOfRemainingPods = podutil.CountAllPodsWithMatchingLabel(client, namespace, map[string]string{"app": "azuredisk-volume-tester"}, "")
+		if numberOfRemainingPods <= 0 || time.Since(startScaleIn) > 40*time.Minute {
+			break
+		}
+		timePassed := time.Since(startScaleIn)
+		switch {
+		case timePassed <= 1*time.Minute:
+			oneMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
+		case timePassed <= 2*time.Minute:
+			twoMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
+		case timePassed <= 3*time.Minute:
+			threeMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
+		case timePassed <= 5*time.Minute:
+			fiveMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
+		case timePassed <= 10*time.Minute:
+			tenMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
+		}
+		time.Sleep(30 * time.Second)
+	}
+	scaleInPodsCompleted := time.Since(startScaleIn).Minutes()
+
+	for {
+		numberOfRemainingVolumeAtts := volutil.CountAllVolumeAttachments(client)
+		if numberOfRemainingVolumeAtts <= 0 || time.Since(startScaleIn) > 60*time.Minute {
+			break
+		}
+	}
+	scaleInCompleted := time.Since(startScaleIn).Minutes()
+
+	klog.Infof("Scaling out test completed in %f minutes.", scaleOutCompleted)
+	klog.Infof("1 min: %d pods are ready", oneMinPodsRunning)
+	klog.Infof("2 min: %d pods are ready", twoMinPodsRunning)
+	klog.Infof("3 min: %d pods are ready", threeMinPodsRunning)
+	klog.Infof("5 min: %d pods are ready", fiveMinPodsRunning)
+	klog.Infof("10 min: %d pods are ready", tenMinPodsRunning)
+	klog.Infof("Total number of pods in Running state: %d", numberOfRunningPods)
+
+	klog.Infof("Scaling in test completed in %f minutes.", scaleInPodsCompleted)
+	klog.Infof("1 min: %d pods are deleted", oneMinPodsDeleted)
+	klog.Infof("2 min: %d pods are deleted", twoMinPodsDeleted)
+	klog.Infof("3 min: %d pods are deleted", threeMinPodsDeleted)
+	klog.Infof("5 min: %d pods are deleted", fiveMinPodsDeleted)
+	klog.Infof("10 min: %d pods are deleted", tenMinPodsDeleted)
+	klog.Infof("Total number of remaining pods: %d", numberOfRemainingPods)
+
+	klog.Infof("Scaling in detach completed in %f minutes.", scaleInCompleted)
+
+}
+
+func (t *PodSchedulingOnFailoverScaleTest) Run(client clientset.Interface, namespace *v1.Namespace, schedulerName string) {
```

Comment:
I think when we test for failover we want to have shared disks with `maxshares > 1` to have replica attachments. We should check for replica creation and since it is hard to create a "real" failover at scale, we might just want to measure the time it takes for pods to get into running state after deletion. Currently, the test doesn't report any latency metrics.  
For "real" failover I was thinking that we might want to cordon half of the nodes in the beginning of the test, create the pods, then cordon the other half of the nodes while uncordoning the first half and only then delete the pods. This will cause the pods to failover to a new node. But with Sung Jun's new changes, I am not sure we will be creating any replica attachments on un-schedulable(cordonded) nodes, so this approach might not be effective.

Reply:
@landreasyan From this [doc](https://docs.microsoft.com/en-us/azure/virtual-machines/disks-shared-enable?tabs=azure-portal),
`For each disk, you can define a maxShares value that represents the maximum number of nodes that can simultaneously share the disk.`

Do we want simultaneous sharing of the disks here?

Reply:
What I mean is, if the maxshares value is not passed, we won't be creating any replica attachments, thus there is no point in testing the failover with the current setup. We need to pass at least `maxshares=2` to create one replica per disk to test the failover with 1 replica attachment.

Reply:
We have decided to address this issue in a separate PR: https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/issues/321